docs/cardano-monads.rst

```diff
@@ -0,0 +1,467 @@
+
+* Compile-time performance. No type families, no instance search tricks. We've
+  been bitten by this before. Keep it simple.
+
+* Run-time performance. We want a flattened runtime representation for
+  Reader-isomorphic effects.
+
+* Run-time configurability. It's fine to keep track of effects themselves in the
+  types, but the choice of an implementation must be delegated to terms. For
+  instance, choosing between a real DB (RocksDB) and pure DB must be possible
+  with a CLI option.
+
+* Predictability. It must be easy manipulate effects in a predictable manner,
+  without fearing that ``runReaderT`` will affect anything but ``MonadReader``.
+
+Proposed Solutions
+------------------
+
+The ReaderT Pattern
+~~~~~~~~~~~~~~~~~~~
+
+Read the full article: https://www.fpcomplete.com/blog/2017/06/readert-design-pattern
+
+The general sentiment of the article is to have a single `Env` data type
+(basically, what we have with our modes), and use `ReaderT Env IO` exclusively.
+
+The general sentiment is true: `StateT`, `WriterT`, and `ExceptT` are almost
+never what you actually want. `StateT` kills the story for concurrency and makes
+it harder to deal with exceptions, `WriterT` is simply broken performance-wise
+(and its CPS-ed version is `StateT` with the same caveats), and `ExceptT` adds
+one more way to throw exceptions (so you need one more way to catch them).
+
+For pure code (without `IO` or `MonadIO`), use `StateT`, `WriterT`, and
+`ExceptT`. In other cases, avoid them.
+
+This aspect of the `ReaderT` pattern makes a lot of sense. Monad transformers
+are used to model effects in pure code, and in `IO` you don't have a choice but
+to have the actual effects themselves (there's mutable state, there are
+exceptions). Mixing them is unhelpful.
+
+Now, let's consider the global `Env` type. The issue with it is associated
+boilerplate of `Has` classes. In case we want to extend the environment to
+introduce local effects, there's a high cost in boilerplate. This can be already
+observed in our codebase.
+
+Point-by-point rundown:
+
+* Flexibility: LOW. Extension of environment requires declaring a new type,
+  writing `Has` instances for it, and potentially monad instances (although the
+  approach recommends avoiding monad classes in favor of `Has` classes, not
+  everything is under our control and there are classes by external libraries).
+  There's no way to partially run effects -- one needs to supply the entire
+  `ReaderT` environment up front.
+
+* Extensibility: MODERATE. Defining new effects boils down to declaring a method
+  record and a `Has` class for it. Then, for each field in the method record we
+  must define a function that extracts this method using the `Has` class and
+  runs it. Boilerplate-heavy, yes, but at least effects don't need to know about
+  each other.
+
+* Ease of use: MODERATE. Beginners can quickly grasp the basic concepts, but
+  there are pitfalls when it comes to lifting method records (see section on
```

Comment:
Why isn't it possible to live with mutable state and configuration options in `Env` exclusively? So that we don't have any horrible `a -> (s -> ScaryT m b) -> GhastlyT m c` methods inside.

Reply:
I don't understand the question, sorry.

Reply:
`Env` fields are of two types: 0-arg functions (`TVar Mem` or `Config`) and n-arg functions (as i've illustrated in the comment above). Why can't we get rid of type (2) completely?

Reply:
Are you talking about method dictionaries in a `ReaderT` environment? Well, you have to pass them one way or another

Reply:
Another -- as a typeclass method, yes. That's what I'm talking about. 

Method dictionaries with `n` methods provide much more variety of combinations rather than instances. Most of the time you don't need dicts.